text/0227-separate-semantic-conventions.md

```diff
@@ -0,0 +1,156 @@
+  create instrumentation compliant with semantic conventions.
+  - `resource/` - Contents of `{specification}/resource/semantic_conventions`
+  - `trace/` - Contents of `{specification}/trace/semantic_conventions`
+  - `metrics/` - Contents of `{specification}/metrics/semantic_conventions`
+  - `logs/`- Contents of `{specification}/logs/semantic_conventions`
+  - `schemas/` - A new location for [Telemetry Schemas](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/schemas/README.md)
+    to live. This directory will be hosted at
+    `https://opentelemetry.io/semantic_conventions/schemas/`
+
+Existing semantic conventions in the specification would be marked as
+deprecated, with documentation denoting the move.
+
+Additionally, if the semantic conventions eventually move to domain-specific
+director structure (e.g. `docs/{domain}/README.md`, with trace, metrics, events
+in the same file), then this can be refactored in the new repository, preserving
+git history.
+
+There will also be the following exceptions in the specification:
+
+- Semantic conventions used to implement API/SDK details will be fully specified
+  and will not be allowed to change in the Semantic Convention directory.
+  - Error/Exception handling will remain in the specification.
+  - SDK configuration interaction w/ semantic convention will remain in the
+    specification. Specifically `service.name`.
+- The specification may elevate some semantic conventions as necessary for
+  compatibility requirements, e.g. `service.instance.id` and
+  [Prometheus Compatibility](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md).
+
+These exceptions exist because:
+
+- Stable portions of the specification already rely on these conventions
+- These conventions are required to implement an SDK today.
+
+As such, the Specification should define the absolute minimum of reserved or
+required attribute names and their interaction with the SDK.
+
+## Internal details
+
+The following process would be used to ensure semantic conventions are
+seamlessly moved to their new location.
+
+- A new repository `open-telemetry/semantic-conventions` will be constructed with
+  the proposed format and necessary `Makefile` / tooling.
+- A moratorium will be placed on Semantic Convention PRs to the specififcation
+  repository. (Caveat that PRs related to this proposal would be allowed).
+- All semantic conventions in the Specification will be copied to the new
+  repository.
+- Semantic conventions in the Specification will be marked as deprecated with
+  links to the new location.
+- The Specification will be updated to require "special" conventions, like
+  `service.name` and configuration interaction.
+- Instrumentation authors will update their `SchemaURL` to
+   `https://opentelemetry.io/semantic_conventions/schemas/{semconv_version}`
```

Comment:
Do we specify anywhere that convention version numbers follow semver rules? I can't find it in our spec. If we don't do that then why is it implied that 1.x is stable?

Reply:
Here's the docs:

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/schemas/README.md#schema-version-number

```
Version numbers use the [ordering rules](https://semver.org/#spec-item-11) defined by semver 2.0 specification. See how ordering is used in the [Order of Transformations](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/schemas/file_format_v1.1.0.md#order-of-transformations). Other than the ordering rules the schema version numbers do not carry any other semantic meaning.
```

So you're right there is no semantic meaning of breaking dependency.  However, effectively that has crippled SchemaURLs right now.  The only mechanism we have to denote "breaking" is changing the Schema family -> i.e. the URL.   OpenTelemetry is *required* to use the same URL in the current experimental specification, therefore OTEL cannot ever break.

I can update this proposal to include some fixes to how SchemaURL interacts.  I *am* suggested the current specification for SchemaURL is problematic.

Reply:
If we need a way to denote breaking changes let's work on that. If we agree that the semver will be how we want to denote that then I will agree with you that we need to move back to 0.x.

However, I am not sure that that is the right solution to denote the breaking change. I think going to 0.x is unnecessarily disruptive. I think we may want to find a different solution that is less disruptive.

To understand what the solution should be I think we need to first explain what the problem is. I am not sure I understand the problem and I do not see it described anywhere. So to me the first question to answer is: why is it necessary to denote which semantic conventions are stable vs unstable and why does it need to be possible to understand by looking at the Schema URL? What is the expectation from the consumers of Schema URL? Why do the consumers care?

>I am suggested the current specification for SchemaURL is problematic.

That's entirely possible :-)

Reply:
I understand you find this proposal disruptive.  I think moving semconv to a new location is *inherently* disruptive.  What this proposal does is try to embrace the distruptive in a way that's beneficial to community.  Specifically by:

1. Making it clear the disruption is happening.
2. Allowing Language SiGs to migrate-at-their-own pace.

If you think there's a way to do this switch with less disruption, I'd like to see the proposal.  I want to avoid the situation w/ Spec changes + Language SiGs are held "hostage" by a large forced migration to the new repository.

Reply:
>I think moving semconv to a new location is inherently disruptive.

Moving the content to a different repository does not need to be disruptive from schema publishing perspective. We can continue the current schema family with zero disruption.

>What this proposal does is try to embrace the distruptive in a way that's beneficial to community. Specifically by:
>Making it clear the disruption is happening.
>Allowing Language SiGs to migrate-at-their-own pace.

I think the disruption is avoidable, so we don't necessarily need to embrace it. :-)

I see 2 elements in this change:
1. Moving the source of truth for conventions to a different repository decouples them from the specification and allows conventions to be freed from overly strict processes we rightfully have in place for the spec repo and instead adopt their own processes. It also allows finer granularity in permission control. This will allow semantic convention contributors to move faster and in my opinion is a good idea and is a win for everyone.
2. Schemas that describe semantic conventions are published from this spec repo, which is currently the source of truth for conventions. We plan to move the source of truth to a different physical repository. This is fine and I do not see any reason for people who _consume_ schemas. They don't need to know which github repository contains the original schema files. Today the schemas are published at https://opentelemetry.io/schemas and they can continue being published at the same place after the original schema files are moved from spec repo to a semconv repo.


>If you think there's a way to do this switch with less disruption, I'd like to see the proposal. 

Here is I think what we can do:
1. Copy the existing semconv content and schema files from the spec repo to a new repo.
1. Create new teams and introduce any new processes wanted on the new repo.
1. Delete convention yaml files and schema yaml files from spec repo to prevent accidental usage in the future. 
1. Language SIGs can begin using the yaml files from the new repo next time they need update their generated semconvs. The [semconv generator](https://github.com/open-telemetry/build-tools/tree/main/semantic-conventions) should work fine with the new repo. SIGs can do this at their pace.
1. Add a notice in convention md files in the spec repo saying that semconv is moved.
1. Change opentelemetry.io/schemas publishing process to consume schema files from the new repo.
1. Continue publishing schemas within the same schema family, without resetting version numbers.

Let me know if you want me to elaborate on the process.

>I want to avoid the situation w/ Spec changes + Language SiGs are held "hostage" by a large forced migration to the new repository.

I may be missing something but I don't see the large forced migration that you are referring to. Can you please clarify?

Reply:
So before (1) we need to place a moratorium on semconv in the Specification so that no changes CAN occur during the transition.

Regarding (3) we cannot delete the yaml files until language sigs have moved their code generation, others we've possibly broken their builds or process.  Also, deleting this would break our existing checks on markdown + semconv.

Additionally, there should be a clear signal from language SiGs that the migration has completed so we know when we could remove semconv from the latest branches in the specification.

Reply:
>So before (1) we need to place a moratorium on semconv in the Specification so that no changes CAN occur during the transition.

Sounds good. We should aim to execute this plan as quickly as possible to minimize the time period when we cannot change/add conventions.

>Regarding (3) we cannot delete the yaml files until language sigs have moved their code generation, others we've possibly broken their builds or process. 

From what I understand code generation is used only when a new version of semconv is introduced (new version of spec today). Language SIGs will not need to use the code generator because we will not be making new releases until the transition to the new repo is complete.

We can double check this with language SIGs to make sure what I am saying is correct. If any of the language sigs uses the code generator for any other purpose (e.g. as part of regular builds somehow) then we can do what you suggest and delay the deletion of exiting yaml files until no-one depends on them anymore.

The code generator currently requires the path to the yaml files to be specified as a [command line parameter](https://github.com/open-telemetry/build-tools/tree/main/semantic-conventions#code-generator). It is not hard-coded to the spec repo in any way. So there are no changes in the generator itself and for language SIGs the change is minimal, they will need to update their processes to reference the new repo (which the code generator expects to be somewhere in a local directory). Here is for example [how it is done in Go](https://github.com/open-telemetry/opentelemetry-go/blob/main/RELEASING.md#semantic-convention-generation).

I have not checked the other languages, we can work with all maintainers to confirm that they do something similar.

>Also, deleting this would break our existing checks on markdown + semconv.

We should also remove the checks when we delete the yamls. Markdown will be manually marked "deprecated" or "moved" and will be frozen in that state. Checks won't be necessary anymore.

Reply:
> We can double check this with language SIGs to make sure what I am saying is correct. If any of the language sigs uses the code generator for any other purpose (e.g. as part of regular builds somehow) then we can do what you suggest and delay the deletion of exiting yaml files until no-one depends on them anymore.

We rely on the YAML files in the specification itself to ensure all Markdown tables are actually up-to-date with the YAML.  That's the main reason we can't delete them.

> The code generator currently requires the path to the yaml files to be specified as a [command line parameter](https://github.com/open-telemetry/build-tools/tree/main/semantic-conventions#code-generator).

Yes, I'm aware of this.  The automation still hardcodes pulling down the specification repo, so that needs to change.  While it's not a large change, it's still something every langauge/collector will need to do.

> We should also remove the checks when we delete the yamls. Markdown will be manually marked "deprecated" or "moved" and will be frozen in that state. Checks won't be necessary anymore.

That's fair.  I still think we should delay doing this until we're comfortable with the new repository and have fully shifted to it.

Reply:
Sounds good. Let's just aim to make the transition period as short as possible. I can help with execution.

Reply:
Think we're aligned, updated the specification appropriately.